ambry-router/src/test/java/com.github.ambry.router/PutManagerTest.java

```diff
@@ -699,34 +835,124 @@ private void verifyBlob(String blobId, byte[] originalPutContent, HashMap<String
+          if (key != lastKey) {
+            Assert.assertEquals("all chunks except last should be fully filled", chunkSize, dataBlobLength);
+          } else {
+            // If we are here (composite blob), we know that the blob size is non-zero.
+            Assert.assertEquals(
+                "Last chunk should be of non-zero length and equal to the length of the remaining bytes",
+                (originalPutContent.length - 1) % chunkSize + 1, dataBlobLength);
+          }
+          offset += dataBlobLength;
+          notificationSystem.verifyNotification(key.getID(), NotificationBlobType.DataChunk,
+              dataBlobPutRequest.getBlobProperties(), dataBlobPutRequest.getUsermetadata().array());
+        }
+        Assert.assertArrayEquals("Input blob and written blob should be the same", originalPutContent, content);
+      } else {
+        // since offset is accessed from inner class (callback), have to make it final and hence AtomicInteger
+        final AtomicInteger offset = new AtomicInteger(0);
+        for (StoreKey key : dataBlobIds) {
+          PutRequest.ReceivedPutRequest dataBlobPutRequest = deserializePutRequest(serializedRequests.get(key.getID()));
+          int dataBlobLength = (int) dataBlobPutRequest.getBlobSize();
+          InputStream dataBlobStream = dataBlobPutRequest.getBlobStream();
+          byte[] dataBlobContent = Utils.readBytesFromStream(dataBlobStream, dataBlobLength);
+          // decrypt and verify the blob content and user-metadata matches
+          // reason to execute via executor service is that, any exception(and assertion failures) thrown by a different
+          // thread than the main thread will not fail the test. With this approach, future.get() will rethrow Exceptions
+          // and Errors as ExecutionException
+          Future<?> future = executorService.submit(new Thread(
+              new DecryptJob(dataBlobPutRequest.getBlobId(), dataBlobPutRequest.getBlobEncryptionKey().duplicate(),
+                  ByteBuffer.wrap(dataBlobContent), dataBlobPutRequest.getUsermetadata().duplicate(), cryptoService,
+                  kms, new Callback<DecryptJob.DecryptJobResult>() {
+                @Override
+                public void onCompletion(DecryptJob.DecryptJobResult result, Exception exception) {
+                  Assert.assertNull("Exception should not be thrown", exception);
+                  Assert.assertEquals("BlobId mismatch", dataBlobPutRequest.getBlobId(), result.getBlobId());
+                  Assert.assertArrayEquals("UserMetadata mismatch", originalUserMetadata,
+                      result.getDecryptedUserMetadata().array());
+                  int resultSize = result.getDecryptedBlobContent().remaining();
+                  result.getDecryptedBlobContent().get(content, offset.get(), resultSize);
+                  offset.addAndGet(resultSize);
+                  if (key != lastKey) {
```

Comment:
I did try to do that actually. It might seem similar, but there are minor subtleties involved. Will sync up with you offline on this. We can try to see if we can modularize this